sql: index out of range during EXPLAIN (TYPES) of DELETE FROM USING

pkg/ccl/testccl/sqlccl/sqlccl_test.TestExplainRedactDDL [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10723627?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10723627?buildTab=artifacts#/) on release-23.1 @ [f13c021bb97531ac876c2c1438a9d4bfc80d1c32](https://github.com/cockroachdb/cockroach/commits/f13c021bb97531ac876c2c1438a9d4bfc80d1c32):

Fatal error:

```
panic: runtime error: index out of range [2] with length 2 [recovered]
	panic: runtime error: index out of range [2] with length 2 [recovered]
	panic: runtime error: index out of range [2] with length 2 [recovered]
	panic: runtime error: index out of range [2] with length 2 [recovered]
	panic: runtime error: index out of range [2] with length 2

```
Stack: 

```
goroutine 535549 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc00adec000, {0x73f7f40, 0xc00c8c6720}, {0x5ac9480?, 0xc00dd77098?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1159 +0x1b4
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1({0x73f7f40, 0xc00c8c6720}, {0xda1bfc9c?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:898 +0x4a
panic({0x5ac9480, 0xc00dd77098})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x407
panic({0x5ac9480, 0xc00dd77098})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x407
panic({0x5ac9480, 0xc00dd77098})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/sql.emitExplain.func1()
	github.com/cockroachdb/cockroach/pkg/sql/explain_plan.go:194 +0x4d
panic({0x5ac9480, 0xc00dd77098})
	GOROOT/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain.(*OutputBuilder).Expr.func1(0xc00305f740, 0x2)
	github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain/output.go:159 +0xc8
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*IndexedVar).Format(0x0?, 0x73b27c0?)
	github.com/cockroachdb/cockroach/pkg/sql/sem/tree/indexed_vars.go:86 +0xa7
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).formatNodeMaybeMarkRedaction(0xc00305f740?, {0x73b27c0?, 0xc0097ba5d0?})
	github.com/cockroachdb/cockroach/pkg/sql/sem/tree/format.go:692 +0x263
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).FormatNode(0xc00305f740, {0x73b27c0, 0xc0097ba5d0})
	github.com/cockroachdb/cockroach/pkg/sql/sem/tree/format.go:440 +0x2d3
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain.(*OutputBuilder).Expr(0xc0048767d0, {0xc003656a80, 0xd}, {0x742e448, 0xc0097ba5d0}, {0xc005ae9d10, 0x2, 0x2})
	github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain/output.go:163 +0x1b3
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain.(*emitter).emitNodeAttributes(0xc004876128, 0xc0097ba500)
	github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain/emit.go:601 +0xa419
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain.Emit.func1(0xc004876128?)
	github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain/emit.go:53 +0x11b
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain.Emit.func1(0xc004876138?)
	github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain/emit.go:57 +0x1a2
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain.Emit(0xc0097baa80, 0xc0048767d0, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain/emit.go:103 +0x8c5
github.com/cockroachdb/cockroach/pkg/sql.emitExplain(0xc0048767d0?, 0xc00549bc80?, {{0xc0013b49f0?, 0xc0013b4a08?}, {0xc0013b49f0?}, 0x0?}, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/explain_plan.go:228 +0xe5
github.com/cockroachdb/cockroach/pkg/sql.(*explainPlanNode).startExec(0xc003edb000, {{0x73f7f40, 0xc00549bbc0}, 0xc003edc800, 0xc00adec620})
	github.com/cockroachdb/cockroach/pkg/sql/explain_plan.go:132 +0x75d
github.com/cockroachdb/cockroach/pkg/sql.startExec.func2({0xc004876b28?, 0x464904?}, {0x73fa5c0, 0xc003edb000})
	github.com/cockroachdb/cockroach/pkg/sql/plan.go:519 +0x143
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1()
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:112 +0x3e
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal(0xc004876c38, {0x73fa5c0?, 0xc003edb000?}, {0x5e2172c?, 0xc?})
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:299 +0x1523
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc004876c38, {0x73fa5c0, 0xc003edb000})
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:79 +0xf7
github.com/cockroachdb/cockroach/pkg/sql.walkPlan(...)
	github.com/cockroachdb/cockroach/pkg/sql/walk.go:43
github.com/cockroachdb/cockroach/pkg/sql.startExec({{0x73f7f40?, 0xc00549bbc0?}, 0xc003edc800?, 0xc00adec620?}, {0x73fa5c0, 0xc003edb000})
	github.com/cockroachdb/cockroach/pkg/sql/plan.go:522 +0x119
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start(0xc00b43a840, {0x73f7f40, 0xc00549bb60})
	github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:174 +0x113
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Init(0xc004737800, {0x73f7f40, 0xc00549bb00})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:183 +0xf3
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Init(0xc003edb500, {0x73f7f40?, 0xc00549bb00?})
	github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:67 +0x4f
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).init(0x557c020?)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:94 +0x2c
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc003edb540?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x62
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Init(0xc002f23ef0, {0x73f7f40?, 0xc00549bb00?})
	github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:103 +0x7f
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).init.func1()
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:245 +0x2e
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0x2?)
	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x62
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).init(0x73f7e98?, {0x73f7f40?, 0xc00549bb00?})
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:244 +0x4f
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).Run(0xc00ed088c0, {0x73f7e98?, 0xc003edb440?})
	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:278 +0xe7
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run(0xc00b01e300, {0x73f7e98, 0xc003edb440}, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:325 +0x20d
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc001220000, {0x73f7f40, 0xc00549b7d0}, 0xc007242580, 0xc001924b00, 0xc0097bad80, 0xc005262700, 0xc00adeca58, 0xc00371f9d0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:902 +0xb93
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0x0?, {0x73f7f40, 0xc00549b7d0}, 0xc004877b58?, 0xc007242580, 0xc003edb000?, {{0x73fa5c0?, 0xc003edb000?}, 0x0?}, 0xc005262700, ...)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1913 +0x205
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll.func3(0xc00adec620, 0x0?, {0x73f7f40, 0xc00549b7d0}, 0xc00549b800?, 0x0?, 0xc004877b88?)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1645 +0xcb
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll(0xc007242580?, {0x73f7f40?, 0xc00549b7d0}, 0xc00adeca58, 0xc007242580, 0xc00adec620, 0xc005262700, 0x0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1648 +0x245
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc00adec000, {0x73f7f40?, 0xc00549b7d0}, 0xc00adec620, 0xc00549b7d0?, {0x7feec7291d80?, 0xc003346728}, 0xc003346728?, 0xc00259c398)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2103 +0x5eb
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc00adec000, {0x73f7f40, 0xc00cc49830}, 0xc00adec620, {0x7feec7291d80, 0xc003346728})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1656 +0x11b2
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc00adec000, {0x73f7f40, 0xc00cc49830}, {{0x7423780, 0xc0010e8640}, {0x0, 0x0, 0x0}, {0xc00377b416, 0x171}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:964 +0x3a51
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x73f7f40?, 0xc00cc49650?})
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:142 +0xbc
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0x73f7f40?, {0x73f7f40?, 0xc00cc49650?}, {0x7423780?, 0xc0010e8640?}, 0x64?, 0x0?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2986 +0x330
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc00adec000, {0x73f7f40, 0xc00cc49650}, {{0x7423780, 0xc0010e8640}, {0x0, 0x0, 0x0}, {0xc00377b416, 0x171}, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:141 +0x4dd
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0xc004879a78, 0xc00adec000, 0xc0048795d0, 0xc0048796a0, 0xc0048796e0, 0xc0048796c0, 0xc0048796b0, 0xc0048796d0)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2181 +0x319
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc00adec000)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2186 +0x1167
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc00adec000, {0x73f7f40, 0xc00c8c6720}, 0xc00242a2e0?, 0xc0047fbda0?, 0xc00992e950?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2103 +0x216
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc00242a280?, {0x73f7f40?, 0xc00c8c6720?}, {0x5ded873?}, 0x3?, 0xc002640410?)
	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:900 +0xe6
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1()
	github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:310 +0x585
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
	github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:219 +0x22a
```
<details><summary>Log preceding fatal error</summary>
<p>

```
*   | 	GOROOT/src/runtime/panic.go:113
*   | github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain.(*OutputBuilder).Expr.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain/output.go:159
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*IndexedVar).Format
*   | 	github.com/cockroachdb/cockroach/pkg/sql/sem/tree/indexed_vars.go:86
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).formatNodeMaybeMarkRedaction
*   | 	github.com/cockroachdb/cockroach/pkg/sql/sem/tree/format.go:692
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).FormatNode
*   | 	github.com/cockroachdb/cockroach/pkg/sql/sem/tree/format.go:440
*   | github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain.(*OutputBuilder).Expr
*   | 	github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain/output.go:163
*   | github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain.(*emitter).emitNodeAttributes
*   | 	github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain/emit.go:601
*   | github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain.Emit.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain/emit.go:53
*   | github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain.Emit.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain/emit.go:57
*   | github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain.Emit
*   | 	github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain/emit.go:103
*   | github.com/cockroachdb/cockroach/pkg/sql.emitExplain
*   | 	github.com/cockroachdb/cockroach/pkg/sql/explain_plan.go:228
*   | github.com/cockroachdb/cockroach/pkg/sql.(*explainPlanNode).startExec
*   | 	github.com/cockroachdb/cockroach/pkg/sql/explain_plan.go:132
*   | github.com/cockroachdb/cockroach/pkg/sql.startExec.func2
*   | 	github.com/cockroachdb/cockroach/pkg/sql/plan.go:519
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/walk.go:112
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal
*   | 	github.com/cockroachdb/cockroach/pkg/sql/walk.go:299
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit
*   | 	github.com/cockroachdb/cockroach/pkg/sql/walk.go:79
*   | github.com/cockroachdb/cockroach/pkg/sql.walkPlan
*   | 	github.com/cockroachdb/cockroach/pkg/sql/walk.go:43
*   | github.com/cockroachdb/cockroach/pkg/sql.startExec
*   | 	github.com/cockroachdb/cockroach/pkg/sql/plan.go:522
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start
*   | 	github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:174
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Init
*   | 	github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:183
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Init
*   | 	github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:67
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).init
*   | 	github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:94
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError
*   | 	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Init
*   | 	github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:103
* Wraps: (4) runtime error: index out of range [2] with length 2
* Error types: (1) *withstack.withStack (2) *safedetails.withSafeDetails (3) *withstack.withStack (4) runtime.boundsError
*
```
</p>
</details>
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExplainRedactDDL.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-29207

--- a/pkg/sql/opt/exec/execbuilder/testdata/delete
+++ b/pkg/sql/opt/exec/execbuilder/testdata/delete
@@ -337,3 +337,47 @@ vectorized: true
               estimated row count: 990 (missing stats)
               table: xyz@xyz_y_idx
               spans: /1-/1000 /2001-/3000
+
+# Testcase for issue 105803.
+
+statement ok
+CREATE TABLE b (b INT PRIMARY KEY)
+
+query T
+EXPLAIN (TYPES) DELETE FROM a USING b WHERE b > 1 RETURNING a, b, NULL
+----
+distribution: local
+vectorized: true
+·
+• render
+│ columns: (a int, b int, "?column?" unknown)
+│ render ?column?: (NULL)[unknown]
+│ render a: (a)[int]
+│ render b: (b)[int]
+│
+└── • delete
+    │ columns: (a int, b int)
+    │ estimated row count: 1,000 (missing stats)
+    │ from: a
+    │ auto commit
+    │
+    └── • distinct
+        │ columns: (a int, b int)
+        │ estimated row count: 1,000 (missing stats)
+        │ distinct on: a
+        │
+        └── • cross join (inner)
+            │ columns: (a int, b int)
+            │ estimated row count: 333,333 (missing stats)
+            │
+            ├── • scan
+            │     columns: (a int)
+            │     estimated row count: 1,000 (missing stats)
+            │     table: a@a_pkey
+            │     spans: FULL SCAN
+            │
+            └── • scan
+                  columns: (b int)
+                  estimated row count: 333 (missing stats)
+                  table: b@b_pkey
+                  spans: /2-

--- a/pkg/sql/opt/exec/explain/result_columns.go
+++ b/pkg/sql/opt/exec/explain/result_columns.go
@@ -150,7 +150,10 @@ func getResultColumns(
 
 	case deleteOp:
 		a := args.(*deleteArgs)
-		return tableColumns(a.Table, a.ReturnCols), nil
+		return appendColumns(
+			tableColumns(a.Table, a.ReturnCols),
+			a.Passthrough...,
+		), nil
 
 	case opaqueOp:
 		if args.(*opaqueArgs).Metadata != nil {

